text/0812-tracked-built-ins.md

```diff
@@ -0,0 +1,287 @@
+} from 'tracked-built-ins';
+```
+
+For example, we could make a scoreboard component that keeps score for an
+arbitrary number of players, and keep track of the score for each player using
+tracked version of JavaScript [Map][global-objects-map].
+
+**Note:** In this and the following examples, I am using `<template>` from
+[RFC #779][rfc-0779].
+
+```js
+// app/components/scoreboard.js
+import Component from '@glimmer/component';
+import { action } from '@ember/object';
+import { TrackedMap } from 'tracked-built-ins';
+
+const PLAYERS = [
+  { name: 'Zoey' },
+  { name: 'Tomster' },
+];
+
+export default class Scoreboard extends Component {
+  // Create a map from player -> score, with each player's score starting at 0
+  scores = new TrackedMap(PLAYERS.map(p => [p, 0]));
+
+  @action
+  incrementScore(player) {
+    let currentScore = this.scores.get(player);
+
+    this.scores.set(player, currentScore + 1);
+  }
+
+  <template>
+    {{#each-in this.scores |player score|}}
+      <div>
+        {{player.name}}: {{score}}
+    
+        <button type="button" {{on "click" (fn this.incrementScore player)}}>
+          +
+        </button>
+      </div>
+    {{/each-in}}
+  </template>
+}
+```
+
+Lets lake a look into another example involving a shopping cart, where we keep track
+of the items added to cart using tracked version of JavaScript's [Array][global-objects-array].
+
+```js
+// app/components/scoreboard.js
+import Component from '@glimmer/component';
+import { action } from '@ember/object';
+import { TrackedArray } from 'tracked-built-ins';
+
+export default class Cart extends Component {
+  items = new TrackedArray([
+    { productName: 'Jeans' },
+    { productName: 'T-shirt' },
+  ]);
+
+  @action
+  addPants(item) {
+    this.items.push(item);
+  }
+
+  <template>
+    {{#each this.items as |item|}}
+      <div>
+        {{item.productName}}
+      </div>
+    {{/each}}
+  
+    <button type="button" {{on "click" (fn this.addPants @pants)}}>
+      Add Pants
+    </button>
+  </template>
+}
+```
+
+## Drawbacks
+
+> Why should we *not* do this? Please consider the impact on teaching Ember,
+on the integration of this feature with other existing and planned features,
+on the impact of the API churn on existing apps, etc.
+
+> There are tradeoffs to choosing any path, please attempt to identify them here.
+
+## Alternatives
+
+- `tracked-built-ins` can be added only to the `app` blueprint and not to the `addon` blueprint.
+
+- Do nothing and recommend using re-setting the property to track data structure changes.
+
+  The tradeoffs of this approach described in "Motivation" section of this RFC.
+
+- Don't add `tracked-built-ins` to the blueprint and only update Ember Guides.
+
+  Ember.js is advertised as "batteries included" framework with complete out-of-the-box experience.
+  Asking users to install additional dependencies to achieve relatively common problem contradicts
+  with that paradigm.
+
+## Unresolved questions
+
```

Comment:
Rereading this morning, I think we need another unresolved question here, which is:

```suggestion

- What, if any, substantive changes to the APIs presented by `tracked-built-ins` should we make?

```

For example (and I *strongly* suspect from previous discussions that other folks on Framework Core will agree with me on this), I think that while the decisions made around [what to support and what not to support from the actual built-in APIs](https://github.com/tracked-tools/tracked-built-ins/blob/9111be44d781448993c170fb68b59fb8a69667c2/README.md#usage) were perfectly reasonable, as we promote them to official status we probably want to revisit them. In particular:

- The `new TrackedArray()` and `new TrackedObject()` constructors differ from the `new Array()` and `new Object()` constructors; should we align them? For `Array`, it’s just a weird and confusing API, but aligning it with the built-in API is probably still fine: that way we just teach it as matching the built-in, however weird that is. For `new TrackedObject()`, though, we would have to mutate the original object, which seems weird.

- What about the omitted features? Can we make those “just work” by delegating to the class they are tracked versions of?